### ISSUE TITLE
[FR] - Log reason when node switches tip to a fork

### ISSUE THREAD

**Issue by TerminadaPool** [IntersectMBO/cardano-node#6014]
**Internal/External**
*External*:   Stake pool operator

**Area**
*Other*:   Logging

**Describe the feature you'd like**
Please log the reason why the node preferred another fork

**Describe alternatives you've considered**
The independent tool ```cncli``` provides logging of block data including the block VRF value into a sqlite database.  This database can be queried for insight into whether the likely reason the node preferred another fork was because the other fork terminal block had a lower block VRF and the tie break rule was applied.

**Additional context / screenshots**
The node knows the reason for switching tip to a competing fork but this reason isn't logged.  Rather the logs only confirm that the tip was switched.  But, was the reason due to the competing fork being longer?  Or was it due to the other fork's terminal block having a lower VRF causing the tie-break rule to be applied?  Or did the node prefer another fork some other reason?

This information would be really helpful when trying to identify what caused your own block to not be adopted.

**Comment by TerminadaPool** [comment]
Please stay as an open feature request.

**Comment by rphair** [comment]
Yes, please: if this is considered impractical to add I would like to hear why.  Our SPO operation believes this is important for quality assurance purposes.

**Comment by rphair** [comment]
@carbolymer what kind of engagement do the developers expect for this issue not to be "stale"?  There won't be anything new to report until someone the relevant team responds.

This log info has been long & vitally requested for diagnostic purposes and the need for it is never going to go away.  Those who need it most would be small stake pools with infrequent blocks... and therefore needed more in the near future if & when the `K` PoS parameter increases (as is constantly being discussed in governance conversations).

If this issue keeps being dismissed, the implication is that QA isn't very important for stake pools and/or chain density isn't that important for Cardano itself.  I don't think either of these things are true and I can state from SPO experience dating back to the beginning of the Shelley period that implementing this feature (which we arguably should have had from day 1) would be helpful for SPOs and node infrastructure operators at all levels of endeavour.

**Comment by carbolymer** [comment]
There was a slack discussion about this issue some time ago. 
@jasagredo @amesgen would you mind sharing main takeaways here?

**Comment by amesgen** [comment]
The trace messages already contain enough detail to deduce the reason why a node changes its chain. Concretely, whenever we change our selection, we trace the so called "`SelectView`" of the tip of both the old and the new chain. This `SelectView` contains exactly the information that is needed in order to determine whether a chain is preferable to another chain. See [`PraosChainSelectView`](https://ouroboros-consensus.cardano.intersectmbo.org/haddocks/ouroboros-consensus-protocol/Ouroboros-Consensus-Protocol-Praos-Common.html#t:PraosChainSelectView) for more details and assorted documentation.

In Conway, the chain order works like this, via a lexicographical combinationi of the following criteria, when deciding whether to switch from the current chain to a candidate chain:

 1. compare chain length, preferring larger values
 2. when (same slot and same issuer) compare operational certificate issue number, preferring larger values
 3. when (slots differ by at most 5) compare VRF, preferring smaller values

By comparing the old and new `SelectView`, one can see which of these three rules was responsible for the chain switch:

---

Let's look at an example:

```json
{
  "at": "2024-12-13T13:50:58.295305594Z",
  "ns": "ChainDB.AddBlockEvent.AddedToCurrentChain",
  "data": {
    "kind": "AddedToCurrentChain",
    "newTipSelectView": {
      "chainLength": 11214658,
      "issueNo": 16,
      "issuerHash": "8e0bb9b126acd8e65a023602377d9a7a6f15af394d1f7bdd852182b8",
      "kind": "PraosChainSelectView",
      "slotNo": 142531567,
      "tieBreakVRF": "f8f31aabb555124392485cbd7ba02bded4e4eac8980fc36fde0c9a21d7cf2d51c6fd4ddda8113153639db862781dafacbdd77da4e072452a330c32509506762e"
    },
    "newtip": "17d5b0f21e86503b41851640e9558610ebe9b7a62c96a08873eb4d0c7961a460@142531567",
    "oldTipSelectView": {
      "chainLength": 11214657,
      "issueNo": 20,
      "issuerHash": "c63dab6d780a74cbae2a27696c9723f55b3092b2bd001256df03827f",
      "kind": "PraosChainSelectView",
      "slotNo": 142531554,
      "tieBreakVRF": "905ff37db665e86d60d93dabc92ec86124b1a22165e4968facca226e790d9b7dd3f60502fa84dcd30d95ac28a9278e5e72b8b2eb451759ab59c92fc65909a2b4"
    }
  },
  "sev": "Notice",
  "thread": "76",
  "host": "ramify"
}
```

Here, we can see that the `chainLength` of `newTipSelectView` is larger than the one of the `chainLength` of the `oldTipSelectView`, so the chain switch happened due to rule 1.

If these would be equal, one would have to look at rule 2 next (and then rule 3) to see whether it is responsible.

---

For convenience/simplicity, we could further enrich the trace message to directly include eg a `reason` field that says something like `"LongerChain"` or `"BetterTieBreaker"`.

cc @mkarg as you might have already started with this.

**Comment by TerminadaPool** [comment]
That sort of log information would absolutely suit my needs.  I wasn't aware we could get the "tieBreakVRF" values logged ordinarily.

What do I need to configure in my node config.json file so that such information is logged in my systemd journal?

**Comment by amesgen** [comment]
In both the old and new tracing system, you should see this when you enable JSON output and use a severity of at least `Notice` for `ChainDB.AddBlockEvent.AddedToCurrentChain`/`ChainDB.AddBlockEvent.SwitchedToAFork`.

**Comment by mgmeier** [comment]
Thanks @amesgen for summarizing the discussion we had (BTW: needed to edit your post to change my GH handle to @mgmeier - the correct one). I'll add some more

1. As mentioned the above trace messages / tracers need to be enabled, and severity filtering should be configured include messages of serverity `Notice`.
2. In the new system, you'd additionally need to configure the trace messages to be more verbose, by setting their detail level to at least `DDetailed`.
3. We have it on our list to add a convenience field to the trace message, stating a reason for switching a fork (should that happen). Would this make sense to provide at normal detail level, without knowing the underlying detailed criteria (like chain length, opcert no and VRF)?
4. Point 3 will likely happen for the new tracing system only, as the old system will be maintenance-only starting with Node release 10.2 - and not receive additional features.

**Comment by amesgen** [comment]
> BTW: needed to edit your post to change my GH handle to @mgmeier - the correct one

Thanks for catching that!

> 2. In the new system, you'd additionally need to configure the trace messages to be more verbose, by setting their detail level to at least `DDetailed`.

In the code, it looks like `{new,old}TipSelectView` is traced even with the default level of detail:
https://github.com/IntersectMBO/cardano-node/blob/1774c93f9174022f359fab3bfe81d9427e32634a/cardano-node/src/Cardano/Node/Tracing/Tracers/ChainDB.hs#L511-L518

**Comment by mgmeier** [comment]
> In the code, it looks like {new,old}TipSelectView is traced even with the default level of detail
Aplogies, yes, the info will be traced at default detail level.
I was looking at a local branch where I was toying around with the settings :-|

**Comment by rphair** [comment]
> In both the old and new tracing system, you should see this **when you enable JSON output**

@amesgen @mgmeier our pool has 4+ years of homebrewed automation driven from text-based output (`"ViewMode": "SimpleView"`).

Does https://github.com/IntersectMBO/cardano-node/issues/6014#issuecomment-2574907814 mean that `AddedToCurrentChain` and `SwitchedToAFork` are not visible from `SimpleView`, and that therefore we must change to JSON logging to see these events?

**Comment by mgmeier** [comment]
Which tracing system are you using (i.e., what is your config value for `UseTraceDispatcher`)?

**Comment by mgmeier** [comment]
Then it is the old system, as currently it defaults to `false`.
As far as I can see, the info for `{new,old}TipSelectView` is indeed only available for JSON logging.

However, here's an idea about how to scrape that info without breaking existing automation. In the node config, you could set up an additional scribe to a file, having `"scFormat": "ScJson"`. You would then need to create a script, possibly based on `grep` and `jq`, that monitors the trace messages (=JSON objects) going to that file. When you encounter a message indicating switching to a fork, the script could inspect its data and send an event to the systemd journal, to merge with the rest of the logging data in text-based format. If you configure log rotation or somesuch for that file scribe, the script would need to take that into account somehow as well.

Would that work for you somehow?

**Comment by rphair** [comment]
yes, thanks @mgmeier - to log these events we'll fork off another log as you suggest. 🙏 If whatever changes are contemplated here produce effects in `SimpleView` I trust we will be able to find out about them here or in a linked thread.

**Comment by rphair** [comment]
@TerminadaPool - as the OP, how do you feel about requiring a 2nd JSON format log output to make these events visible?  And is this issue resolved as you see it? (the marking `Stale` as a foot-in-the-door to closing this FR doesn't seem like a proper resolution to me: seems like it should be either closed "Won't Fix" or given a label that keeps it open until the feature is provided).

**Comment by rphair** [comment]
@github-actions I am running out of things to say here.

**Comment by rphair** [comment]
Hello @github-actions!

**Comment by rphair** [comment]
@github-actions @mgmeier I do think this issue is still relevant and still hope that it can be catalogued in such a way that we don't have to confirm this every 30 days to avoid the issue becoming "stale" and disappearing.

**Comment by TerminadaPool** [comment]
I don't actually separately configure an extra .json file output for logging but instead use the normal journal output.  I would appreciate a normal log message stating the reason for why the node switched tip.

**Comment by rphair** [comment]
🎶

**Comment by TerminadaPool** [comment]
In node version 10.4.1 I have configured MachineFormat for Stdout backend in TraceOptions:
```
"TraceOptions": {
    "": {
      "backends": [
        "Stdout MachineFormat",
```
MachineFormat provides information like this for each block adopted:
```
Jun 18 18:50:40 relay2 cardano-node[2570632]: {"at":"2025-06-18T08:50:40.790719758Z","ns":"ChainDB.AddBlockEvent.AddedToCurrentChain","data":{"kind":"AddedToCurrentChain","newTipSelectView":"chainLength":12012087,"issueNo":14,"issuerHash":"8d2fef46df5f9b2fcb64ebad739f27062b4f38c04df1a916fa216752","kind":"PraosChainSelectView","slotNo":158670349,"tieBreakVRF":"42e07a46dc4aeccb949caaef8c3341392ae0744a361fd854123d314f37723702950c80615201b90b18276239375a1f02ebe189e1243260347d9cc039941804bb"},"newtip":"6b01316c4536981cb767ff63521aa3d857a83f271e040a89d85471ea5fbc097f@158670349","oldTipSelectView":"chainLength":12012086,"issueNo":26,"issuerHash":"e985b5698de1386439e62b5b7ab2cc9f814b91b4a1fef5970f38694d","kind":"PraosChainSelectView","slotNo":158670327,"tieBreakVRF":"f5d16073b6bad50b1517830e70ef7a252ceada181cd7b4e7f6f35908487e7be14f7859a7f7c2d8997c29e9b23a9a844dd69dd3c03dbc72491877ea550cbd0247"}},"sev":"Notice","thread":"79","host":"relay2"}
```
It is a bit of a pain to read this json format in the journal, but you can see the tieBreakVRF value.  When there is a "switched tip to new fork" event you can see that the node followed the fork with the lower tieBreakVRF.

Unfortunately you don't get the tieBreakVRF values in the logs if you use the human readable formats: HumanFormatColoured or HumanFormatUncoloured.  But at least the spelling of "coloured" is correct. :)

**Comment by rphair** [comment]
🔥

**Comment by rphair** [comment]
🎶 ("on hold" music... going on for many months now)

**Comment by rphair** [comment]
👽

**Comment by rphair** [comment]
🎄

**Comment by rphair** [comment]
@amesgen @carbolymer @mgmeier - can we have a revisitation with some proper commentary (please also disarming the 30-day kill switch) on this request... now that the last 30-day interval (lately, to repeatedly ignore it) has presented the "mainnet incident" for which chain selection was operationally crucial for chain survival and disaster recovery?  cc @TerminadaPool

**Comment by mgmeier** [comment]
Performance & Tracing will pick this issue up.

**Comment by rphair** [comment]
@mgmeier @jutaro how can this issue be automatically be marked `Stale` when it's been assigned to somebody & that label has been deliberately removed?  How can it be that the CI here really puts such an issue on track for automatic closure? 🤔

**Comment by rphair** [comment]
🤖

**Comment by jasagredo** [comment]
This feature was implemented in Consensus in https://github.com/IntersectMBO/ouroboros-consensus/pull/1826 and in Cardano-node in https://github.com/IntersectMBO/cardano-node/pull/6469.

Once we have confirmation that the new information is being logged correctly, we can finally close this issue.

@mgmeier can I ask you to confirm this is the case?

**Comment by rphair** [comment]
thanks @jasagredo @mgmeier — along with this confirmation, please (for the sake of the SPO community) post a snippet of the log message that we should be looking for... in the new tracing system as well as the old tracing system if applicable. 🙏

**Comment by mgmeier** [comment]
The old tracing system **does not** support this new message field.

In the new tracing system, it's logged under the `ChainDB.AddBlockEvent.SwitchedToAFork` namespace.
The message is of severity `Notice`, so it's only filtered out if you set your config to minimum severity `Warning` or higher.

**Comment by mgmeier** [comment]
This is an example text output:
```
[2026-04-14 18:57:00.6869Z][0979accfcade:ChainDB.AddBlockEvent.SwitchedToAFork](Notice,25) Switched to a fork, new tip: 6ca9f1fc17c029a120f662531f1dd93148cdddd35d4c386897fd37be754339db at slot 942
Reason: candidate is longer (At (BlockNo 29)) than our selection (At (BlockNo 28))
```

**Comment by mgmeier** [comment]
This is an example JSON output (I linted it, it's actually a one-liner):
```
{
	"at": "2026-04-14T20:28:13.589952121Z",
	"ns": "ChainDB.AddBlockEvent.SwitchedToAFork",
	"data": {
		"kind": "TraceAddBlockEvent.SwitchedToAFork",
		"newSuffixSelectView": {
			"blockNo": 23,
			"issueNo": 0,
			"issuerHash": "87569d82e26ad45fb249a162ff5a71f71808746286ae2ecad3ccd34c",
			"kind": "PraosTiebreakerView",
			"slotNo": 705,
			"tieBreakVRF": "6143ed8f0f398121e66cddf0b7e7f1e30daa94492261ba2709d1c286c02edcb2b2a07f994ce9b58235483a667ccc236a82039a286c07a2f97705f5010f04ff10",
			"weightBoost": 0
		},
		"newtip": "6af05875b569a6c3510dae170b1c31507795ce76e75b94e83e32c6026392bb3d@705",
		"oldSuffixSelectView": {
			"blockNo": 22,
			"issueNo": 0,
			"issuerHash": "842e4a3eb479cec07a5df650f913d6931485f9e466f21247d3cb1fd3",
			"kind": "PraosTiebreakerView",
			"slotNo": 902,
			"tieBreakVRF": "f5198cc0b901b80735fea02da3accc292a14371ae2c0f7c6670161dc3c025b9f78eceeb71e6e2067d2274749733aad6e9e7a6eb024925eda96bddd7ae96884f3",
			"weightBoost": 0
		},
		"reason": {
			"candidate": "At (BlockNo 23)",
			"our": "At (BlockNo 22)",
			"reason": "Longer"
		}
	},
	"sev": "Notice",
	"thread": "25",
	"host": "10d0916dc292"
}
```
